4Science/DSpace
# Port "identifiers during submission" to DSpace CRIS 7 REST

## References
This is a port of the simple submission step (now section) for DSpace 6.3

* 6.3 module: https://github.com/the-library-code/identifiers-enduring-submission/tree/dspace-6.3-addition
* dspace-angular UI component PR: https://github.com/4Science/dspace-angular/pull/5
* Related to [REST Contract - submission section types ](https://github.com/4Science/Rest7Contract/blob/dspace-cris-7/submissionsection-types.md) - a new type will need to be added

## Description
A new submission section class populates submission section data with identifiers that will be / have been minted for an item (in the case of DOIs, they are pre-minted before returning the section data) so that submitters can update their PDFs or other references with DOIs, Handles etc before completing the submission.

This component was contributed by The Library Code with the support of Technische Universität Hamburg.

## Instructions for Reviewers
**NOTE**: The Integration Test for this work is still a work-in-progress. The step works in real usage, but for some reason the REST / API test environment is not adding the new section to test items submissions. Any guidance is appreciated.

**BUG**: There is a strange race condition when an item filter is used with the DOI provider, where the subscription to sectionData means that when the item goes from "no match" to "match" on the DOI filter, it reserves OK, but the next time the item is updated the DOI will be minted again, and receive a new URI. This is still being analyzed - there are probably improvements that can be made to limit / throttle the calls back to the section data?

List of changes in this PR:
* New step and integration test: `ShowIdentifiersStep` and `SubmissionShowIdentifiersRestIT`
* New simple bean (data model): `DataIdentifiers`
* Test submission configuration updated for steps and submission processes, to include the new `identifiers` section

To test:
* Add a new item (right now, the step is just configured for publication form)
* Check that the handle is displayed in the first section accordion
* If DOIs are enabled, check that the DOI is displayed in the first section accordion

## Checklist
_This checklist provides a reminder of what we are going to look for when reviewing your PR. You need not complete this checklist prior to creating your PR (draft PRs are always welcome). If you are unsure about an item in the checklist, don't hesitate to ask. We're here to help!_
